aclk/agent_cloud_link.c

```diff
@@ -169,6 +165,10 @@ static int create_private_key()
     private_key[bytes_read] = 0;
     debug(D_ACLK, "Claimed agent loaded private key len=%zu bytes", bytes_read);
     fclose(f);
+    if (bytes_read == 0) {
```

Comment:
Shouldn't this be `bytes_read != statbuf.st_size` or something?

Reply:
> Shouldn't this be `bytes_read != statbuf.st_size` or something?

It detects (probably) that you are trying to use `statbuf.st_size` after `lstat` succeeds but there is no guarantee that once you hit `fopen` it will be the same file (in size and all).

Reply:
No - the check is to detect that the file was empty not that the size changed.

Reply:
If you have a 64 byte file and you read 4 bytes then it is an error, no?

Reply:
What if the file changed between the time of the lstat and the time of the read (which is the whole point of the TOCTOU check in Coverity).

Reply:
The problem is the call `fread(private_key, 1, statbuf.st_size, f);` which has not been fixed.

If `statbuf.st_size` is smaller than the real length then you don't read the whole value.

Moreover, if `statbuf.st_size` is exactly right, which is more likely, but you fail to read the whole file (because `bytes_read` was cut short) then it's also invalid.

You need to set the correct length for the `fread()` call and check `if (bytes_read != length)`. The "or something" comment is probably solved by doing something like:
https://stackoverflow.com/questions/238603/how-can-i-get-a-files-size-in-c

Reply:
This is where we've done this before: https://github.com/netdata/netdata/blob/8fe7485a600f60be9ac20af923a61d516cb0335e/health/health.c#L52

Reply:
But you are trying to check the validity of the file by considering if the number of bytes read in is the expected length. This does not tell us that the file or its contents are valid. Attempting to load the key into openssl is what tells us that the contents were valid. The piece of code that was moved was simply to check that we do not try to read a zero-length file. It was not necessary, but was there as an early bail for a condition that is clearly wrong.

Fixing the TOCTOU has been done. The fix that you are trying to put into the code is not valid, or necessary. The reason that I moved the check rather than eliminate it completely is because I cannot tell whether calling the openssl function with a zero-length string is problematic. Calling it with invalid data will cause a failure to initialise the key.

Reply:
Calling `fread()` with the length `statbuf.st_size` isn't TOCTOU?

Reply:
I don't think so. There are two scenarios:
1. An attacker appends something to the file. It is not read by the call. No damage results.
2. An attacker trunctates the file. The call returns the number of bytes read, our memory is not corrupted and the key fails to validate when we call openssl.

Are you thinking of another case?

Reply:
The problem is not security, the problem is that it's a bug. I explained above what can go wrong. The user impact would be that an operation would fail when it wouldn't need to. If coverity is happy with this change then you can ignore this. In the health code this is how @thiagoftsm fixed the coverity defect though and I believe this is the correct approach.

Reply:
Can you explain a scenario in which you think the bug will be triggered and what will occur?

Reply:
The race condition is rare to begin with, coverity worries about `statbuf.st_size` being different than the file length during `fopen()`.

It's mostly going to be exactly the same 99,999% of the cases. Especially since I don't believe we're generating variable sizes for keys, although I'm not sure about key size differences.

If somehow the file-system does not atomically update this file, which I consider improbable due to its small size, the scenario is the following:

- real length < `statbuf.st_size`:
  - this is safe because `fopen()` will stop at the end of file.
- real length > `statbuf.st_size`:
  - this will cause the code to read only a part of the key and it's the bug.

Moreover, we have to account for errors in `fopen()` returns fewer bytes than what it asked, but this may be unrelated to the coverity report.

Reply:
If `create_private_key()` indicates a failure then it gets called again, so in the improbable situation that the file (which is 1679 bytes long) is not atomically updated during the claim and it changes size between the two calls... we simply call it again and it is validated correctly. In fact looking at the main loop in the aclk thread I see that we call it one more time after it has succeeded, I would speculate that nobody knows why, possibly for luck? 

The security problem is a potentially a real problem which is why Coverity checks it as it gives an attacker a way to affect a code path inside the code, potentially causing a buffer overwrite. That is not a problem the way that we've written the code so in other places in Netdata the same pattern has been marked as a false positive (if you recall we had a long conversation about the portability of `fstat` last time this happened).

So although technically the piece of code can go wrong locally, it is within a larger context that will stop it being a problem. So I am happy that it will not cause a problem for users. There is a long philosophical discussion that we are not going to have about whether or not it is still a bug. Yes, in general the fseek/ftell is a better way to read the file although there potential for it to go wrong. On Linux it should be fine but not every platform returns the number of bytes from ftell (I can't remember which one doesn't so I don't know if we support it or not).

So if we've finished establishing whether or not the code is broken, can we proceed with reviewing the Coverity fixes as I would like us to get the release out on time...

Reply:
`ftell()` should be standard C, if it's indeed problematic then the existing health code has an issue. If coverity is happy with the code then I'll leave it at that. I didn't request a code change in any case.

Reply:
I'll look up the `ftell` thing then and see if my memory is correct in case we have a problem.